dttools/src/catalog_query.c

```diff
@@ -13,45 +14,147 @@ See the file COPYING for details.
 #include "xxmalloc.h"
 #include "stringtools.h"
 #include "debug.h"
+#include "datagram.h"
+#include "domain_name_cache.h"
+#include "domain_name.h"
+#include "set.h"
+#include "list.h"
 
 struct catalog_query {
 	struct jx *data;
 	struct jx *filter_expr;
 	struct jx_item *current;
 };
 
-struct catalog_query *catalog_query_create(const char *host, int port, struct jx *filter_expr, time_t stoptime)
+struct catalog_host {
+	char *host;
+	char *url;
+	int down;
+};
+
+static struct set *down_hosts = NULL;
+
+/* Given a semicolon delimited list of host:port or host, set the values pointed
+   to by host and port, using the default port if not provided. Return the address
+   of the next hostport in the string, or NULL if there are no more
+*/
+const char *parse_hostlist(const char *hosts, char *host, int *port)
 {
-	if(!host)
-		host = CATALOG_HOST;
-	if(!port)
-		port = CATALOG_PORT;
+	const char *next = strchr(hosts, ';');
+	switch (sscanf(hosts, "%[^:;]:%d", host, port)) {
+	case 1:
+		*port = CATALOG_PORT;
+		break;
+	case 2:
+		break;
+	default:
+		debug(D_DEBUG, "bad host specification: %s", hosts);
+		return NULL;
+		break;
+	}
+	return next ? next + 1 : NULL;
+}
 
-	char *url = string_format("http://%s:%d/query.json", host, port);
-	struct link *link = http_query(url, "GET", stoptime);
-	free(url);
+struct catalog_query *catalog_query_create(const char *hosts, struct jx *filter_expr, time_t stoptime)
```

Comment:
This function is getting a bit hairy, can you factor it into one function that queries one host, and a parent function that works on the whole list?